### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by elkezza**
Adding exception handling, by introducing `ExoscaleAPIException` as base class for all exoscale exceptions,`ExoscaleAPIClientException` for handling client-side errors 4xx and `ExoscaleAPIServerException` for server-side errors 5xx.

Tested them locally and works as expected.

### CODE REVIEW

File: exoscale/api/v2.py [hunk 1]
```diff
@@ -155,6 +156,17 @@ def _call_operation(self, operation_id, parameters=None, body=None):
         response = self.session.request(
             method=op["verb"].upper(), url=url, params=query_params, **json
         )
+
+        # Error handling
+        if response.status_code >= 400 and response.status_code < 500:
```
Comment (brutasse) [inline exoscale/api/v2.py:161]: A separate Exception class for auth-related errors would be useful. It can be raised whenever the API responds with a 403.

**Comment by brutasse** [comment]
Fixes #61

**Review by brutasse — CHANGES_REQUESTED** [review]
Thanks @elkezza. This looks good but it would be useful to have some basic tests:

- mock the HTTP library to simulate HTTP error codes
- make API requests and ensure they raise the appropriate exceptions

This can be done in `tests/test_client.py`. Thanks :+1:

**Comment by elkezza** [comment]
@brutasse thank you for your feedback, I will do accordingly.

File: tests/test_client.py [hunk 2]
```diff
@@ -20,3 +24,54 @@ def test_client_creation():
...
+    )
+    try:
+        client.get_instance(id="85664334-0fd5-47bd-94a1-b4f40b1d2eb7")
+    except ExoscaleAPIClientException as e:
+        assert "Client error 404" in str(e)
+        assert '{"message":"Instance not found"}' in str(e)
+
+    # Mock a 500 server error
+    requests_mock.get(
+        "https://api-ch-gva-2.exoscale.com/v2/500_error",
```
Comment (brutasse) [inline tests/test_client.py:61]: Can you simulate a proper API operation raising a 500? This custom path means you have to resort to using `client.session` below which isn't part of the documented interface.

Comment (elkezza) [reply tests/test_client.py]: I didn't find any use-case to trigger 5xx error, however in the new commit I tried to mock use-case where the end point is disabled.
Please let me know if any change needed.

File: tests/test_client.py [hunk 3]
```diff
@@ -20,3 +24,51 @@ def test_client_creation():
...
+        "https://api-ch-gva-2.exoscale.com/v2/template",
+        status_code=503,
+        text='{"message":"Endpoint template temporarily unavailable"}',
+    )
+    try:
+        client.list_templates()
+    except ExoscaleAPIServerException as e:
+        assert "Server error 503" in str(e)
+    except Exception as e:
+        assert "503 Server Error" in str(e)
```
Comment (brutasse) [inline tests/test_client.py:70]:
```suggestion
```

Not sure why this would be necessary - other than that the adjustment looks good :+1:

Comment (elkezza) [reply tests/test_client.py]: Yes it is redundant, good catch, thank you : ).

Comment (elkezza) [reply tests/test_client.py]: @brutasse after your green light, is more reviewer required, to merge the branch?

File: exoscale/api/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# exceptions.py
+
+
```
Comment (brutasse) [inline exoscale/api/exceptions.py:3]:
```suggestion
```

**Review by brutasse — APPROVED** [review]
:+1: LGTM, you can merge after the final suggested change @elkezza. Thanks!

**Comment by elkezza** [comment]
> 👍 LGTM, you can merge after the final suggested change @elkezza. Thanks!

**Comment by elkezza** [comment]
> > 👍 LGTM, you can merge after the final suggested change @elkezza. Thanks!

ack, thanks!

**Review by pierre-emmanuelJ — COMMENTED** [review]
LGTM thanks 👍

**Pull request merged**